### DISCUSSION TITLE
Editable shapes

### DISCUSSION THREAD

**Discussion by thruflo** [electric-sql/electric#1677] [Feature requests]
At the moment, Shapes have a fixed definition. This is provided as a root table and where clause. For example, you may define a shape as "all projects with IDs in [1, 2, 3, ...]".

Now, imagine that you are invited to a new project. You would like to also subscribe to that project. Currently, this can be achieved by:

1. creating a new shape subscription for the single new project
2. abandoning the previous all projects shape subscription and starting a new one with the new list of project IDs

(1) is not very ergonomic to the developer. (2) is not very efficient from a data point of view. Depending on how you're keeping data in the client, you may need to re-sync lots of data you've already synced.

So, this feature request is to **add the capability edit the definition of an existing shape subscription**.

## Implications

When the shape definition changes, the client needs:

- to get any additional data that's now in the shape streamed to them in the form of inserts and updates
- to be notified about any data that has previously need synced but is no longer in the shape

Clients need to handle inserts, updates and deletions anyway, so this may not require any changes in the client stream processing. However, there may be consistency and integrity implications for clients materialising data into a database or normalised store.

## Implementation

How to implement this requires some thought. Internally, at some level, shape logs need to be immutable. So there's a question about how to marry up a flexible, ergonomic, mutable shape definition interface for developers with efficient, cachable shape logs.

**Comment by plutou-ben** [comment]
This please. 

To start and stop shapes for this are exhausting and put too much load on ElectricSQL sync than required
